---
title: "robseq_example"
output: github_document
date: "2023-11-17"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Importing libraries
```{r}
library(Robseq)
library(edgeR)
library(doParallel)
library(EnhancedVolcano)
```

# Loading Example Data
Loading Colon cancer data
```{r}
load("~/Current Data Path/Colon Cancer.RData") 
```

Extracting Colon cancer gene expression data and metadata
```{r}
features = data$counts
metadata = data$metadata
```

# Snapshot of data
A typical bulk RNA-seq gene expression count data frame looks something like below. Note, the genes should be in the rows and the samples in columns
```{r}
features[1:5, 1:5]
```
A typical metadata data frame looks something like below. Note, in our pipeline the user must include a column labeled as "Exposure" which should be the variable which will be used by Robseq in performing differential expression analysis. If the user has a different label for the treatment/condition or disease status variable then the user should supply that name via "expVar" argument in Robseq
```{r}
metadata[1:5, ]
```

# Preprocessing
A typical preprocessing step is to filter lowly abundant genes. We do so using edgeR's "filterByExpr" function
```{r}
keep.exprs <- filterByExpr(features, group = as.factor(metadata$Exposure))
paste(length(which(!keep.exprs)), ' lowly expressed genes were filtered out', sep = '')
features <- features[keep.exprs, ]
```

# Performing differential expression analysis using Robseq
To perform differential gene expression (DGE) analyses using Robseq please use the code below. Note, if your metadata has only the "Exposure" variable (such as treatment groups, conditions, disease status, and etc.) set the "coVars" argument to "NULL". However, in this working example our metadata contained three covariates such as "post-mortem interval (pmi)", "rna integrity number (rin)" and "age at death" which needed to be adjusted for in our DGE analysis. Therefore, we have supplied this three variables to the "coVars" argument
```{r}
fit <- Robseq::robust.dge(features = features,
                          metadata = metadata,
                          norm.method = "RLE",
                          expVar =  "Exposure",
                          coVars = NULL,
                          parallel = TRUE,
                          ncores = detectCores() - 2,
                          verbose = FALSE)
```

# Obtaining results from Robseq
After performing DGE analysis you can extract the results table in the following manner
```{r}
results <- fit$res
```
The results table from Robseq should look something like the following
```{r}
results[1:5,]
```

# Volcano plot to visualize DGE results
Volcano plot to visualize the DGE results obtained from Robseq
```{r}
  EnhancedVolcano(results,
    lab = results$Genes,
    x = 'log2FC',
    y = 'adjPval')
```
